File: tests/search.test.ts
import { describe, expect, test } from "bun:test";
import type { EmbeddingBackend } from "../src/embeddings/openai.ts";
import { BM25Index } from "../src/index/bm25.ts";
import { VectorIndex } from "../src/index/dense.ts";
import { hybridSearch } from "../src/search.ts";
import type { Chunk } from "../src/types.ts";
import { unitVector } from "./test-helpers.ts";

function chunk(content: string, file = "src/a.ts"): Chunk {
  return {
    content,
    file_path: file,
    start_line: 1,
    end_line: 10,
    language: "typescript",
  };
}

function mockEmbeddings(vectors: Float32Array[]): EmbeddingBackend {
  const dim = vectors[0]?.length ?? 4;
  return {
    model: "mock",
    dimensions: dim,
    async embedDocuments(texts: string[]) {
      return texts.map((text) => {
        const index = Number.parseInt(text, 10);
        if (Number.isNaN(index)) {
          return vectors[0] ?? new Float32Array(dim);
        }
        const vec = vectors[index];
        if (!vec) {
          throw new Error(`Missing mock vector for ${text}`);
        }
        return vec;
      });
    },
    async embedQuery(text: string) {
      const parsed = Number.parseInt(text, 10);
      if (!Number.isNaN(parsed)) {
        const [vec] = await this.embedDocuments([text]);
        if (vec) {
          return vec;
        }
      }
      return vectors[0] ?? new Float32Array(dim);
    },
  };
}

describe("hybridSearch", () => {
  test("semantic-only retrieval returns nearest chunk", async () => {
    const chunks = [chunk("0", "src/auth.ts"), chunk("1", "src/db.ts"), chunk("2", "src/util.ts")];
    const vectors = [unitVector(4, 0, 1), unitVector(4, 1, 1), unitVector(4, 2, 1)];
    const bm25 = new BM25Index();
    bm25.index(chunks.map((c) => [c.content]));
    const semantic = new VectorIndex(vectors);
    const embeddings = mockEmbeddings(vectors);

    const results = await hybridSearch({
      query: "0",
      embeddings,
      semanticIndex: semantic,
      bm25Index: bm25,
      chunks,
      topK: 1,
      alpha: 1,
      rerank: false,
    });

    expect(results).toHaveLength(1);
    expect(results[0]?.chunk.file_path).toBe("src/auth.ts");
  });

  test("BM25-only retrieval ranks lexical match first", async () => {
    const chunks = [
      chunk("database migration schema", "src/db.ts"),
      chunk("auth middleware token", "src/auth.ts"),
    ];
    const vectors = [unitVector(4, 0, 1), unitVector(4, 1, 1)];
    const bm25 = new BM25Index();
    bm25.index(chunks.map((c) => c.content.split(/\s+/)));
    const semantic = new VectorIndex(vectors);
    const embeddings = mockEmbeddings(vectors);

    const results = await hybridSearch({
      query: "auth token",
      embeddings,
      semanticIndex: semantic,
      bm25Index: bm25,
      chunks,
      topK: 1,
      alpha: 0,
      rerank: false,
    });

    expect(results[0]?.chunk.file_path).toBe("src/auth.ts");
  });
});

File: internal/search/search_test.go
package search

import (
	"strconv"
	"testing"

	"github.com/takara-ai/miru-code/internal/index"
	"github.com/takara-ai/miru-code/internal/utils"
)

type mockBackend struct {
	vectors [][]float32
}

func (m mockBackend) ModelName() string {
	return "mock"
}

func (m mockBackend) Dimensions() int {
	if len(m.vectors) == 0 {
		return 0
	}
	return len(m.vectors[0])
}

func (m mockBackend) EmbedDocuments(texts []string) ([][]float32, error) {
	out := make([][]float32, len(texts))
	for i, text := range texts {
		idx, err := strconv.Atoi(text)
		if err != nil {
			out[i] = m.vectors[0]
			continue
		}
		out[i] = m.vectors[idx]
	}
	return out, nil
}

func (m mockBackend) EmbedQuery(text string) ([]float32, error) {
	idx, err := strconv.Atoi(text)
	if err != nil {
		return m.vectors[0], nil
	}
	return m.vectors[idx], nil
}

func searchChunk(content, file string) utils.Chunk {
	return utils.Chunk{Content: content, FilePath: file, StartLine: 1, EndLine: 10, Language: "typescript"}
}

func TestHybridSearchSemanticOnlyRetrievalReturnsNearestChunk(t *testing.T) {
	chunks := []utils.Chunk{
		searchChunk("0", "src/auth.ts"),
		searchChunk("1", "src/db.ts"),
		searchChunk("2", "src/util.ts"),
	}
	vectors := [][]float32{unitVector(4, 0), unitVector(4, 1), unitVector(4, 2)}
	bm25 := index.NewBM25Index()
	bm25.Index([][]string{{"0"}, {"1"}, {"2"}})
	semantic := index.NewVectorIndex(vectors)
	alpha := 1.0

	results, err := HybridSearch(HybridOptions{
		Query: "0", Embeddings: mockBackend{vectors: vectors}, SemanticIndex: semantic,
		BM25Index: bm25, Chunks: chunks, TopK: 1, Alpha: &alpha,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) != 1 || results[0].Chunk.FilePath != "src/auth.ts" {
		t.Fatalf("results = %#v, want auth chunk", results)
	}
}

func TestHybridSearchBM25OnlyRetrievalRanksLexicalMatchFirst(t *testing.T) {
	chunks := []utils.Chunk{
		searchChunk("database migration schema", "src/db.ts"),
		searchChunk("auth middleware token", "src/auth.ts"),
	}
	vectors := [][]float32{unitVector(4, 0), unitVector(4, 1)}
	bm25 := index.NewBM25Index()
	bm25.Index([][]string{{"database", "migration", "schema"}, {"auth", "middleware", "token"}})
	alpha := 0.0

	results, err := HybridSearch(HybridOptions{
		Query: "auth token", Embeddings: mockBackend{vectors: vectors}, SemanticIndex: index.NewVectorIndex(vectors),
		BM25Index: bm25, Chunks: chunks, TopK: 1, Alpha: &alpha,
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(results) == 0 || results[0].Chunk.FilePath != "src/auth.ts" {
		t.Fatalf("results = %#v, want auth chunk", results)
	}
}

func unitVector(dim, activeIndex int) []float32 {
	v := make([]float32, dim)
	v[activeIndex] = 1
	return v
}
